Add Time Series Maintenance Queue

mrtracy opened the pull request:
Adds a new storage queue, the "Time Series Maintenance Queue", which
periodically initiates the pruning of time series data. This queue primarily
utilizes the previously developed pruning utilities in the time series package.

Commit includes two tests that verify that the queue works correctly; one test
that verifies that the queue passes the correct information to the time series
pruning functions, and another that starts a test server and directly verifies
that time series data is pruned by the queue.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9959)

<!-- Reviewable:end -->

Commit:
Add Time Series Maintenance Queue

Adds a new storage queue, the "Time Series Maintenance Queue", which
periodically initiates the pruning of time series data. This queue primarily
utilizes the previously developed pruning utilities in the time series package.

Commit includes two tests that verify that the queue works correctly; one test
that verifies that the queue passes the correct information to the time series
pruning functions, and another that starts a test server and directly verifies
that time series data is pruned by the queue.

Closes #5910

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 7 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2L97fmM1u_5fanXGG:-KU2L97fmM1u_5fanXGH:bnilr80) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/store.go#L500)):*
> ```Go
> 	SQLExecutor sqlutil.InternalExecutor
> 
> 	// TimeSeriesDataStore is an interface used by the store's Time series
> ```

`s/series/Series/g`

---

*[pkg/storage/ts_maintenance_queue.go, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LEWx7hESZgE0fnrD:-KU2LEWx7hESZgE0fnrE:b-wa4kkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L33)):*
> ```Go
> 
> const (
> 	timeSeriesMaintenanceQueueTimerDuration = time.Second
> ```

Any reason for setting this at a second? I would think the normal pacing of the scanner would be sufficient.

---

*[pkg/storage/ts_maintenance_queue.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LPJ5_YENxKc746CE:-KU2LPJ5_YENxKc746CF:bkujuo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L34)):*
> ```Go
> const (
> 	timeSeriesMaintenanceQueueTimerDuration = time.Second
> 	timeSeriesMaintenanceQueueMaxSize       = 200
> ```

Every other queue uses a max size of `100`. Why is this queue different? The value used by other queues hasn't been tuned, but I see no reason for this to be inconsistent right now. 

---

*[pkg/storage/ts_maintenance_queue.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2Lkb7_VORPMl6hHOI:-KU2Lkb7_VORPMl6hHOJ:bg3gyxs) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L66)):*
> ```Go
> // * Local replica is a convenient way to discover the names of time series
> // stored on the cluster; the names are required in order to effectively prune
> // time series without expensive scans.
> ```

`s/scans/distributed scans/g`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-:-KU2M0mYzlDjFbUfwbve:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/storage/store.go, line 500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2KoDabWiWu1VgA_en:-KU2KoDabWiWu1VgA_eo:bd3uwtm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/store.go#L500)):*
> ```Go
> 	SQLExecutor sqlutil.InternalExecutor
> 
> 	// TimeSeriesDataStore is an interface used by the store's Time series
> ```

s/Maint/maint/, s/Que/que/?

---

*[pkg/storage/store.go, line 736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2KwGFp6aNVKnj1K8k:-KU2KwGFp6aNVKnj1K8l:b-mteht0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/store.go#L736)):*
> ```Go
> 		s.consistencyScanner.AddQueues(s.replicaConsistencyQueue)
> 
> 		// Add time series maintenance queue.
> ```

This might be one of the rare cases in which the code really speaks for itself.

---

*[pkg/storage/ts_maintenance_queue.go, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LT-HnusvxKXN0cEi:-KU2LT-HnusvxKXN0cEj:by8u3z4) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L64)):*
> ```Go
> // * The queue naturally distributes the workload across the cluster in
> // proportion to the number of ranges containing time series data.
> // * Local replica is a convenient way to discover the names of time series
> ```

s/Local replica/Access to a local Replica's data/?

---

*[pkg/storage/ts_maintenance_queue.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LG7iV9MfT_0GhpWZ:-KU2LG7iV9MfT_0GhpW_:b-ajtacx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L66)):*
> ```Go
> // * Local replica is a convenient way to discover the names of time series
> // stored on the cluster; the names are required in order to effectively prune
> // time series without expensive scans.
> ```

This looks like the right commit to introduce some of the comments we talked about in the last PR (i.e. why isn't concurrent trimming of the same series by multiple constituent Replicas an issue).

---

*[pkg/storage/ts_maintenance_queue.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LdTuHNK8Ezz6EAZ0:-KU2LdTv7Qb8crLHWxDn:b-yneehb) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L76)):*
> ```Go
> // timeSeriesMaintenanceQueue.
> func newTimeSeriesMaintenanceQueue(
> 	store *Store, db *client.DB, g *gossip.Gossip, tsData TimeSeriesDataStore,
> ```

You pass a `*Store`, so you already have access to `Gossip`. Any particular reason to add it again?

---

*[pkg/storage/ts_maintenance_queue.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2Ls4Uoo1NRTRqynfx:-KU2Ls4VIRfCuxUE6UnW:b-zfv0jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L99)):*
> ```Go
> 
> func (tsmq *timeSeriesMaintenanceQueue) shouldQueue(
> 	now hlc.Timestamp, repl *Replica, sysCfg config.SystemConfig,
> ```

s/now|sysCfg/_/

---

*[pkg/storage/ts_maintenance_queue.go, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2M3WLsqjMtwI8eML-:-KU2M3WLsqjMtwI8eML0:b-wd7bdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L110)):*
> ```Go
> 	desc := repl.Desc()
> 	eng := repl.store.Engine()
> 	return tsmq.tsData.PruneTimeSeries(ctx, eng, desc.StartKey, desc.EndKey, tsmq.db, now)
> ```

I think it would be more prudent to pass a snapshot into `PruneTimeSeries` as you don't write to that engine and you're not planning to.

---

*[pkg/storage/ts_maintenance_queue_test.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2N01lvtQS1Eug37s5:-KU2N01lvtQS1Eug37s6:bia81e2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue_test.go#L172)):*
> ```Go
> 
> 	// Populate time series data into the server. One time series, with one
> 	// datapoints at the current time and one datapoint older than the pruning
> ```

s/points/point/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-:-KU2Nl1xUlbLMuOD3C1c:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/storage/metrics.go, line 218 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2NxiKDr5Dgrb50Pcb:-KU2NxiKDr5Dgrb50Pcc:b-m3uo6n) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/metrics.go#L218)):*
> ```Go
> 	metaTimeSeriesMaintenanceQueuePending = metric.Metadata{Name: "queue.tsmaintenance.pending",
> 		Help: "Number of pending replicas in the time series maintenance queue"}
> 	metaTimeSeriesMaintenanceQueueProcessingNanos = metric.Metadata{Name: "queue.tsmaintenance.processingnanos",
> ```

*If* we expose the others' metrics in the `ui`,  we should add these too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 3 of 9 files reviewed at latest revision, 13 unresolved discussions.

---

*[pkg/storage/metrics.go, line 218 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2NxiKDr5Dgrb50Pcb:-KU3YxV18gSw_KBfd3Gm:b-1kxti) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/metrics.go#L218)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> *If* we expose the others' metrics in the `ui`,  we should add these too.
</details>
They're not currently in the UI.

---

*[pkg/storage/store.go, line 500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2L97fmM1u_5fanXGG:-KU3ZHpZ-6Hn-msJL2sI:becokuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/store.go#L500)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/series/Series/g`
</details>
i'm going all lower case.

---

*[pkg/storage/store.go, line 500 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2KoDabWiWu1VgA_en:-KU3ZNfr64kMCm-4Mvww:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/store.go#L500)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Maint/maint/, s/Que/que/?
</details>
Done.

---

*[pkg/storage/store.go, line 736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2KwGFp6aNVKnj1K8k:-KU3ZXOkTx36_Ty2Lz4r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/store.go#L736)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This might be one of the rare cases in which the code really speaks for itself.
</details>
Done.

---

*[pkg/storage/ts_maintenance_queue.go, line 33 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LEWx7hESZgE0fnrD:-KU3ZhbDi77LZK5fyYgg:b58djdf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L33)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Any reason for setting this at a second? I would think the normal pacing of the scanner would be sufficient.
</details>
Is this faster than the normal pacing?  I thought the normal pacing was greedy, and this was slowing it down significantly. I borrowed this value from the GC Queue, which I believe is the most similar queue.

---

*[pkg/storage/ts_maintenance_queue.go, line 34 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LPJ5_YENxKc746CE:-KU3ZpFu81AAQnMHTaTE:bxs2qcy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L34)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Every other queue uses a max size of `100`. Why is this queue different? The value used by other queues hasn't been tuned, but I see no reason for this to be inconsistent right now. 
</details>
I would swear that this had been set to 200 on the other queues, I guess I am crazy.  Fixed.

---

*[pkg/storage/ts_maintenance_queue.go, line 64 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LT-HnusvxKXN0cEi:-KU3_m1z4k7QM1QCPTKB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L64)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Local replica/Access to a local Replica's data/?
</details>
Done.

---

*[pkg/storage/ts_maintenance_queue.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2Lkb7_VORPMl6hHOI:-KU3_tgvfQ0Nrd3IR-tf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L66)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/scans/distributed scans/g`?
</details>
Done.

---

*[pkg/storage/ts_maintenance_queue.go, line 66 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LG7iV9MfT_0GhpWZ:-KU3eNXKVB8xwN-puQ9p:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L66)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This looks like the right commit to introduce some of the comments we talked about in the last PR (i.e. why isn't concurrent trimming of the same series by multiple constituent Replicas an issue).
</details>
Done.

---

*[pkg/storage/ts_maintenance_queue.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LdTuHNK8Ezz6EAZ0:-KU3_uaBlbUqSsitFRob:b-hax74y) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L76)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You pass a `*Store`, so you already have access to `Gossip`. Any particular reason to add it again?
</details>
Several of the other "newQueue" methods have this same signature (passing store, db and gossip separately); I was originally passing just the store, and switched after seeing the other methods. I do not know why that signature was chosen for the other queues, I assumed it was for testing or something, or perhaps it's not guaranteed that the store actually has a DB or Gossip; in cases like this I just try to match my signatures to the existing signatures.

---

*[pkg/storage/ts_maintenance_queue.go, line 99 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2Ls4Uoo1NRTRqynfx:-KU3b0tyA0EAoEHhLaaN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L99)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/now|sysCfg/_/
</details>
Done.

---

*[pkg/storage/ts_maintenance_queue.go, line 110 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2M3WLsqjMtwI8eML-:-KU3eM5BsjSSctytiDgY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L110)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I think it would be more prudent to pass a snapshot into `PruneTimeSeries` as you don't write to that engine and you're not planning to.
</details>
Done.

---

*[pkg/storage/ts_maintenance_queue_test.go, line 172 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2N01lvtQS1Eug37s5:-KU3fZGguY8rBQ1Vj21R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue_test.go#L172)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/points/point/
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/ts_maintenance_queue.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LdTuHNK8Ezz6EAZ0:-KU3oFHD4QFslW6Uah9u:b-waacij) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L76)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Several of the other "newQueue" methods have this same signature (passing store, db and gossip separately); I was originally passing just the store, and switched after seeing the other methods. I do not know why that signature was chosen for the other queues, I assumed it was for testing or something, or perhaps it's not guaranteed that the store actually has a DB or Gossip; in cases like this I just try to match my signatures to the existing signatures.
</details>
I suppose arguments could be made either way, but I think I'd rather break with a bad precedent than keep following it. Your call though.

---

*[pkg/storage/ts_maintenance_queue.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU3ojJ2ogB7eU2O-dGl:-KU3ojJ2ogB7eU2O-dGm:b-v60h2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0e20cb83a1bdfe57c72e446a30a09461dce415/pkg/storage/ts_maintenance_queue.go#L68)):*
> ```Go
> // effectively prune time series without expensive distributed scans.
> //
> // Data changes executed by this queue are idempotent; it is explicitly safe
> ```

This will rot when you do the actual pruning. I don't want to push this too far, but perhaps mention the RFC and associated discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/ts_maintenance_queue.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU2LdTuHNK8Ezz6EAZ0:-KU3rF7RugjiyinJZruV:b-ynlrw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cf0f46e0f761430f7e10357d7a5ca5b84c2d3e09/pkg/storage/ts_maintenance_queue.go#L76)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I suppose arguments could be made either way, but I think I'd rather break with a bad precedent than keep following it. Your call though.
</details>
I will just leave it as is; in my opinion the actual choice of little consequence, and thus consistency is the more important virtue.

---

*[pkg/storage/ts_maintenance_queue.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU3ojJ2ogB7eU2O-dGl:-KU3qrvcOSD2Y0o_WNtV:b-uwfbdx) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0e20cb83a1bdfe57c72e446a30a09461dce415/pkg/storage/ts_maintenance_queue.go#L68)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This will rot when you do the actual pruning. I don't want to push this too far, but perhaps mention the RFC and associated discussion.
</details>
I don't understand why the comment shouldn't describe what it does currently. I'm not optimistic about getting the roll-up process implemented soon, its a lower priority than the deletion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/ts_maintenance_queue.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU3ojJ2ogB7eU2O-dGl:-KU3uKsa0hyxqFWKUrGv:b-sjmmnj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0e20cb83a1bdfe57c72e446a30a09461dce415/pkg/storage/ts_maintenance_queue.go#L68)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I don't understand why the comment shouldn't describe what it does currently. I'm not optimistic about getting the roll-up process implemented soon, its a lower priority than the deletion.
</details>
Because the two of us had the discussion, the discussion is not in the RFC, and you may not be the one continuing this work.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/ts_maintenance_queue.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU3ojJ2ogB7eU2O-dGl:-KU3w-_S9MFHisQvRFhf:b-u5s9yt) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0e20cb83a1bdfe57c72e446a30a09461dce415/pkg/storage/ts_maintenance_queue.go#L68)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Because the two of us had the discussion, the discussion is not in the RFC, and you may not be the one continuing this work.
</details>
I will tune up the RFC to more directly address the concerns raised and how they were resolved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/storage/ts_maintenance_queue.go, line 68 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959#-KU3ojJ2ogB7eU2O-dGl:-KU3wneWc_8DO-_OZr2c:b-z67qte) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc0e20cb83a1bdfe57c72e446a30a09461dce415/pkg/storage/ts_maintenance_queue.go#L68)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I will tune up the RFC to more directly address the concerns raised and how they were resolved.
</details>
:thumbsup: that works for me, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9959)*
<!-- Sent from Reviewable.io -->
